test/puppetlabs/puppetdb/testutils/log.clj

```diff
@@ -0,0 +1,164 @@
+  ;; Assumes use of logback (i.e. logger supports Levels).
+  `(let [logger# (#'puppetlabs.puppetdb.testutils.log/find-logger ~logger-id)
+         original-level# (.getLevel logger#)]
+     (try
+       (.setLevel logger# (case ~level
+                            :trace Level/TRACE
+                            :debug Level/DEBUG
+                            :info Level/INFO
+                            :warn Level/WARN
+                            :error Level/ERROR
+                            :fatal Level/ERROR))
+       (do ~@body)
+       (finally (.setLevel logger# original-level#)))))
+
+(defn- log-event-listener
+  "Returns a log Appender that will call (listen event) for each log event."
+  [listen]
+  ;; No clue yet if we're supposed to start with a default name.
+  (let [name (atom (str "log-listener-" (kitchensink/uuid)))]
+    (reify
+      Appender
+      (doAppend [this event] (listen event))
+      (getName [this] @name)
+      (setName [this x] (reset! name x))
+      LifeCycle
+      (start [this] true)
+      (stop [this] true))))
+
+(defn- with-additional-log-appenders-fn [logger-id appenders body]
+  (let [logger (find-logger logger-id)]
+    (try
+      (doseq [appender appenders]
+        (.addAppender logger appender))
+      (body)
+      (finally
+        (doseq [appender appenders]
+          (.detachAppender logger appender))))))
+
+(defmacro with-additional-log-appenders
+  "Runs body with the appenders temporarily added to the logger
+  specified by logger-id.  If logger-id is not a class, it is
+  converted via str."
+  [logger-id appenders & body]
+  `(#'puppetlabs.puppetdb.testutils.log/with-additional-log-appenders-fn
+     ~logger-id ~appenders (fn [] ~@body)))
+
+(defn- with-log-appenders-fn [logger-id appenders body]
+  (let [logger (find-logger logger-id)
+        original-appenders (iterator-seq (.iteratorForAppenders logger))]
+    (try
+      (doseq [appender original-appenders]
+        (.detachAppender logger appender))
+      (with-additional-log-appenders-fn logger-id appenders body)
+      (finally
+        (doseq [appender original-appenders]
+          (.addAppender logger appender))))))
+
+(defmacro with-log-appenders
+  "Runs body with the current appenders of the logger specified by
+  logger-id replaced by the specified appenders.  If logger-id is not
+  a class, it is converted via str."
+  [logger-id appenders & body]
+  `(#'puppetlabs.puppetdb.testutils.log/with-log-appenders-fn
+     ~logger-id ~appenders (fn [] ~@body)))
+
+(defmacro with-log-event-listener
+  "Calls (listen event) for each logger-id event produced during the
+  execution of the body.  If logger-id is not a class, it is converted
+  via str."
+  [logger-id listen & body]
+  ;; Specify the root logger via org.slf4j.Logger/ROOT_LOGGER_NAME.
+  `(with-additional-log-appenders ~logger-id
+     [(doto (#'puppetlabs.puppetdb.testutils.log/log-event-listener ~listen)
+        .start)]
```

Comment:
It shouldn't matter in this case since our appender's stop method just returns true, but perhaps still good practice; I'll add that.